### DISCUSSION TITLE
CloudKit Support

### DISCUSSION THREAD

**Discussion by Mcrich23** [skiptools/skip#113] [Ideas]
I would love to see support for CloudKit! I know it's probably difficult with how CloudKit operates on javascript, but it should be possible and would be so incredibly helpful.

**Comment by marcprux** [comment]
Thanks for the suggestion. CloudKit is under consideration, since it would be great to add support for the same nice cloud APIs on the Android side, just like we do with our [Firestore](https://skip.tools/docs/modules/skip-firebase/) support. But, as you mentioned, there is no native Android support, so we'd be stuck backing it with a JavaScript bridge, which includes limitations [like](https://cdn.apple-cloudkit.com/cloudkit-catalog/):

> For security reasons, a mobile web view must launch the Apple sign-in page in a native browser in order to use iCloud authentication.

I know there have been some experiments with CK on Android, but to my knowledge there aren't any other cross-platform development tools that have good CloudKit support on both iOS and Android. I'd love to be problem wrong, though – or else maybe it will change sometime soon.

**Reply by aehlke** [reply]
Personally it's enough to have CloudKit sync on iOS and simply not available for Android users (I use CloudKit only for syncing my offline client-side DB with other devices, and not for the public database or other more fringe features)

**Comment by Mcrich23** [comment]
As long as we can pass it back, I don't see what the huge issue is? It's just like any other sign in...

**Comment by marcprux** [comment]
Which parts of CloudKit would you be interesting in using? The `CKDatabase`/`CKContainer`/`CKRecord` API or something like `CKSyncEngine` or `NSPersistentCloudKitContainer`?

It might be possible to support some of these using our existing JavaScriptCore engine [skip-script](https://skip.tools/docs/modules/skip-script/), but some parts of it (like live updates) could be a challenge.

> It's just like any other sign in...

Yes, but I expect it isn't all that common for an Android user to have an Apple ID. But I could be wrong about that.

**Comment by Mcrich23** [comment]
You are correct that most android users do not have apple ids. However, they would have to sign up for a service anyway to use their db in a user containerized way, so why not cloudkit. I would be interested in `CKDatabase`, `CKContainer`, `CKRecord`, and `NSPersistentCloudKitContainer`.
